### DISCUSSION TITLE
NVIDIA CUDA works on macOS <= 10.13, but homebrew supports macOS >= 10.14

### DISCUSSION THREAD

**Discussion by webbp** [Homebrew/discussions#1506] [Tap maintenance and Homebrew development]
An upgraded Mac Pro 5,1 is still a respectable machine for creative professionals and machine learning. Mine is from 2009, but is a 12-core 3.06GHz Xeon with 96gb ram and a water-cooled NVIDIA GTX 980 Ti. However, I suspect that I will soon have to switch to Linux on this machine. Why? NVIDIA Toolkit, which is the backend of nearly all modern machine learning software, only works on macOS <= 10.13.6. (This is, of course, Apple and NVIDIA's fault, not Homebrew's.)

This issue affects any Mac user with a powerful NVIDIA card, meaning it is unlikely to affect web software engineers, but it strongly affects many creative professionals and data scientists who use machine learning.

The lack of binary packages is already problematic because a lot of machine learning software takes many hours to build. But more worrisome is that Homebrew formulas will be updated to the point of no longer building on 10.13.6. When that happens, I must either:
1. hold back my local homebrew-core formulas, losing access to newer versions and features in machine learning software, or
2. create a branch in my local homebrew-core to patch 10.13.6 breakages as they come, or
3. switch to Linux, which is ok, just kind of a PITA because then I have to restart to use pro graphics and music apps

Does anyone else have or forsee this become a problem? Got any ideas or solutions I haven't considered? E.g., maybe there is already a community 10.13.6 branch or fork or bottle build server?

Another potential source of hope on the horizon is Vulkan (as opposed to CUDA), which is gradually more supported in machine learning software, supports both NVIDIA and AMD GPUs, and runs on Mac using MoltenVK, a Vulkan implementation using Apple's Metal graphics and compute framework.

**Comment by tjluoma** [comment]
I think an easier solution would be to switch from Homebrew to [MacPorts](https://www.macports.org) which, generally speaking, has better support for older versions of macOS. 

Its homepage says “targeting mainly macOS Mojave v10.14 and later” but it is not a hard-and-fast rule as it is in Homebrew.

I'm happily running MacPorts on some older Macs which cannot be upgraded to newer versions of macOS.
